prov/usnic/src/usdf_wait.c

```diff
@@ -0,0 +1,338 @@
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/* This needs to be included for usdf.h */
+#include "fi.h"
+#include "fi_enosys.h"
+#include <fi_util.h>
+
+#include "usdf.h"
+#include "usdf_cq.h"
+#include "usdf_wait.h"
+
+
+/* Necessary to support top-of-file struct declarations. */
+static int usdf_wait_wait(struct fid_wait *wait_fid, int timeout);
+static int usdf_wait_close(struct fid *waitset);
+static int usdf_wait_control(struct fid *fid, int command, void *arg);
+
+static struct fi_ops_wait usdf_wait_ops = {
+	.size = sizeof(struct fi_ops_wait),
+	.wait = usdf_wait_wait,
+};
+
+static struct fi_ops usdf_wait_fi_ops = {
+	.size = sizeof(struct fi_ops),
+	.close = usdf_wait_close,
+	.bind = fi_no_bind,
+	.control = usdf_wait_control,
+	.ops_open = fi_no_ops_open
+};
+
+static int usdf_wait_trywait(struct fid *fwait)
+{
+	struct usdf_wait *wait;
+	struct dlist_entry *item;
+	struct fid_list_entry *entry;
+	int ret = FI_SUCCESS;
+
+	wait = wait_fidtou(fwait);
+
+	dlist_foreach(&wait->list, item) {
+		entry = container_of(item, struct fid_list_entry, entry);
+
+		switch (entry->fid->fclass) {
+		case FI_CLASS_EQ:
+			continue;
+		case FI_CLASS_CQ:
+			ret = usdf_cq_trywait(entry->fid);
+			if (ret)
+				return ret;
+			break;
+		default:
+			USDF_DBG_SYS(FABRIC, "invalid fid %p\n", entry->fid);
+			return -FI_EINVAL;
+		}
+	}
+
+	return ret;
+}
+
+int usdf_trywait(struct fid_fabric *fabric, struct fid **fids, int count)
+{
+	size_t i;
+	int ret;
+
+	for (i = 0; i < count; i++) {
+		assert(fids[i]);
+
+		switch (fids[i]->fclass) {
+		case FI_CLASS_EQ:
+			continue;
+		case FI_CLASS_CQ:
+			ret = usdf_cq_trywait(fids[i]);
+			break;
+		case FI_CLASS_WAIT:
+			ret = usdf_wait_trywait(fids[i]);
+			break;
+		default:
+			USDF_DBG_SYS(FABRIC, "invalid fid\n");
+			return -FI_EINVAL;
+		}
+
+		if (ret)
+			return ret;
+	}
+
+	return FI_SUCCESS;
+}
+
+/* Since a domain hasn't been opened at the time of wait object creation, open a
+ * device temporarily to check for the group interrupt capability.
+ */
+static int usdf_wait_check_support(struct usdf_fabric *fabric_priv)
+{
+	struct usd_open_params params = {
+		.flags = UOPF_SKIP_PD_ALLOC,
+		.cmd_fd = -1,
+		.context = NULL
+	};
+	struct usd_device *dev;
+	int ret;
+
+	ret = usd_open_with_params(fabric_priv->fab_dev_attrs->uda_devname,
+			&params, &dev);
+	if (ret) {
+		USDF_DBG_SYS(FABRIC,
+				"opening device to check fd support failed.\n");
+		return ret;
+	}
+
+	if (!usd_get_cap(dev, USD_CAP_GRP_INTR)) {
+		USDF_WARN_SYS(FABRIC, "FD request invalid.\n");
+		USDF_WARN_SYS(FABRIC, "group interrupts not supported.\n");
+		return -FI_EOPNOTSUPP;
```

Comment:
You should still close the device, even if group interrupts are not supported.

Also, it would be nice to cache whether group interrupts are not supported so that you only have to check the device once per process (it's not going to change per device, nor is it going to change for the life of the process).